core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -390,14 +404,18 @@ class BrokerMetadataListener(
   }
 
   class GetImageRecordsEvent(future: CompletableFuture[util.List[ApiMessageAndVersion]])
-      extends EventQueue.FailureLoggingEvent(log) with Consumer[util.List[ApiMessageAndVersion]] {
-    val records = new util.ArrayList[ApiMessageAndVersion]()
-    override def accept(batch: util.List[ApiMessageAndVersion]): Unit = {
-      records.addAll(batch)
-    }
-
+      extends EventQueue.FailureLoggingEvent(log) {
     override def run(): Unit = {
-      _image.write(this)
+      val records = new util.ArrayList[ApiMessageAndVersion]
+      val writer = new RecordListWriter(records)
+      val options = new ImageWriterOptions.Builder().
+        setMetadataVersion(_image.features().metadataVersion()).
```

Comment:
This comment also applies to it used in the snapshotter.

Outside of tests it looks like this option is always set to metadata version in the metadata image. What do you think about changing the API so that setting the metadata version is optional. If the metadata version is not set then `MetadataImage.write` will set the metadata version to the version contained in the image?

Reply:
let's defer this to later. there could be some complexity in doing the fallback.